### PR TITLE
break cycles in /recipe/

### LINKED ISSUE #1878: Arcs Circular Dependencies
There are probably more, but these ones were quickly found:

madge --circular .
Processed 1311 files (20.4s) (105 warnings)

✖ Found 12 circular dependencies!

1) runtime/recipe/type-checker.js > runtime/ts-build/type.js
2) runtime/ts-build/type.js > runtime/ts-build/schema.js > runtime/entity.js
3) runtime/ts-build/type.js > runtime/ts-build/schema.js
4) runtime/recipe/type-checker.js > runtime/ts-build/type.js > runtime/ts-build/shape.js
5) runtime/ts-build/type.js > runtime/ts-build/shape.js
6) runtime/ts-build/type.js > runtime/tuple-fields.js
7) runtime/ts-build/type.js > runtime/type-variable.js
8) runtime/manifest.js > runtime/ts-build/storage/storage-provider-factory.js > runtime/ts-build/storage/synthetic-storage.js
9) runtime/handle.js > runtime/ts-build/reference.js
10) runtime/arc.js > runtime/recipe-index.js
11) runtime/ts-build/hosted-slot-consumer.js > runtime/ts-build/hosted-slot-context.js
12) shell/components/data-explorer.js > shell/components/data-item.js

### PR DESCRIPTION
**Pull request by mmandlis**
Fixes #1878

### CODE REVIEW

File: src/runtime/recipe/util.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+ * @license
+ * Copyright (c) 2020 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {Recipe} from './recipe.js';
```
Comment (mmandlis) [inline src/runtime/recipe/util.ts]: The name seems too generic, but recipe-util is already taken...
This is not the best way to determine whether `recipe` is a `RequireSection`, but importing `RequireSection` causes a cycle (tried to move it to its own file, but then it's a cycle between that and recipe). 
Should I add a TODO to find a nicer way?

Comment (shaper) [reply src/runtime/recipe/util.ts]: Sure, a TODO sounds fine. Maybe we can just set a boolean on the recipe when we create it? Just dreaming without looking at code.

File: tools/sigh.ts [hunk 2]
```diff
@@ -537,12 +537,14 @@ async function cycles(args: string[]): Promise<boolean> {
...
     sighLog('');
+  } else if (res.length === 0) {
+    sighLog(`CONGRATS! No circular dependencies found!!!\n`);
   }
 
   // This should only go down!
   // Please adjust this number down when you remove cycles.
   // https://github.com/PolymerLabs/arcs/issues/1878
-  const CURRENT_NUMBER_OF_CYCLES = 6;
+  const CURRENT_NUMBER_OF_CYCLES = 0;
```
Comment (shaper) [inline tools/sigh.ts]: Amazing! We must try building it internally.

**Comment by shans** [comment]
Please check that the same number of tests pass/fail with the NG stack active before landing this. Just switch the flag in flags.ts from false to true and run all the tests - if the same number are passing with and without this patch then all is good :)

**Comment by mmandlis** [comment]
> Please check that the same number of tests pass/fail with the NG stack active before landing this. Just switch the flag in flags.ts from false to true and run all the tests - if the same number are passing with and without this patch then all is good :)

Good point, thanks, I just checked now - same number of tests pass and fail (94 fail), will merge this PR shortly.

File: src/runtime/recipe/util.ts (hunk 1 above)
Comment (mmandlis) [reply src/runtime/recipe/util.ts]: added a todo (waiting for tests to pass and merging...)

**Pull request merged**